core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java

```diff
@@ -0,0 +1,1331 @@
+        }
+
+        return Int128.valueOf(remainder);
+    }
+
+    private static void dividePositives(long dividendHigh, long dividendLow, int dividendScaleFactor, long divisorHigh, long divisorLow, int divisorScaleFactor, long[] quotient, long[] remainder)
+    {
+        if (Int128.compare(divisorLow, divisorHigh, 0, 0) == 0) {
+            throwDivisionByZeroException();
+        }
+
+        // to fit 128b * 128b * 32b unsigned multiplication
+        int[] dividend = new int[NUMBER_OF_INTS * 2 + 1];
+        dividend[0] = lowInt(dividendLow);
+        dividend[1] = highInt(dividendLow);
+        dividend[2] = lowInt(dividendHigh);
+        dividend[3] = highInt(dividendHigh);
+
+        if (dividendScaleFactor > 0) {
+            shiftLeftBy5Destructive(dividend, dividendScaleFactor);
+            shiftLeftMultiPrecision(dividend, NUMBER_OF_INTS * 2, dividendScaleFactor);
+        }
+
+        int[] divisor = new int[NUMBER_OF_INTS * 2];
+        divisor[0] = lowInt(divisorLow);
+        divisor[1] = highInt(divisorLow);
+        divisor[2] = lowInt(divisorHigh);
+        divisor[3] = highInt(divisorHigh);
+
+        if (divisorScaleFactor > 0) {
+            shiftLeftBy5Destructive(divisor, divisorScaleFactor);
+            shiftLeftMultiPrecision(divisor, NUMBER_OF_INTS * 2, divisorScaleFactor);
+        }
+
+        int[] multiPrecisionQuotient = new int[NUMBER_OF_INTS * 2];
+        divideUnsignedMultiPrecision(dividend, divisor, multiPrecisionQuotient);
+
+        pack(multiPrecisionQuotient, quotient);
+        pack(dividend, remainder);
+    }
+
+    /**
+     * Value must have a length of 8
+     */
+    private static void shiftLeftBy5Destructive(int[] value, int shift)
+    {
+        if (shift <= MAX_POWER_OF_FIVE_INT) {
+            multiply256Destructive(value, POWERS_OF_FIVES_INT[shift]);
+        }
+        else if (shift < MAX_POWER_OF_TEN_LONG) {
+            multiply256Destructive(value, POWERS_OF_FIVE_LONG[shift]);
+        }
+        else {
+            multiply256Destructive(value, POWERS_OF_FIVE[shift]);
+        }
+    }
+
+    /**
+     * Divides mutableDividend / mutable divisor
+     * Places quotient in first argument and reminder in first argument
+     */
+    private static void divideUnsignedMultiPrecision(int[] dividend, int[] divisor, int[] quotient)
+    {
+        checkArgument(dividend.length == NUMBER_OF_INTS * 2 + 1);
+        checkArgument(divisor.length == NUMBER_OF_INTS * 2);
+        checkArgument(quotient.length == NUMBER_OF_INTS * 2);
+
+        int divisorLength = digitsInIntegerBase(divisor);
+        int dividendLength = digitsInIntegerBase(dividend);
+
+        if (dividendLength < divisorLength) {
+            return;
+        }
+
+        if (divisorLength == 1) {
+            int remainder = divideUnsignedMultiPrecision(dividend, dividendLength, divisor[0]);
+            checkState(dividend[dividend.length - 1] == 0);
+            arraycopy(dividend, 0, quotient, 0, quotient.length);
```

Comment:
If quotient length is constant you can make a static copy method

Reply:
Preexisting